File: app/src/main/java/net/xaethos/trackernotifier/fragments/NotificationsFragment.java
package net.xaethos.trackernotifier.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import net.xaethos.trackernotifier.R;
import net.xaethos.trackernotifier.adapters.NotificationsAdapter;
import net.xaethos.trackernotifier.adapters.NotificationsDataSource;
import net.xaethos.trackernotifier.adapters.NotificationsDividerDecorator;
import net.xaethos.trackernotifier.api.TrackerClient;
import net.xaethos.trackernotifier.models.Notification;
import net.xaethos.trackernotifier.utils.ViewUtils;

import java.util.List;

import rx.Observable;
import rx.Subscriber;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subscriptions.MultipleAssignmentSubscription;

public class NotificationsFragment extends Fragment {

    TrackerClient mApiClient;

    SwipeRefreshLayout mRefreshView;
    View mEmptyView;

    NotificationsAdapter mAdapter;

    private MultipleAssignmentSubscription mDataSubscription;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        mApiClient = TrackerClient.getInstance(getContext());
        mAdapter = NotificationsAdapter.create();

        mDataSubscription = new MultipleAssignmentSubscription();
        refresh();
    }

    @Override
    public void onDestroy() {
        mDataSubscription.unsubscribe();
        super.onDestroy();
    }

    @Override
    public View onCreateView(
            LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_notifications, container, false);
        Context context = getContext();

        RecyclerView recyclerView = (RecyclerView) root.findViewById(android.R.id.list);
        recyclerView.setLayoutManager(new LinearLayoutManager(context));
        recyclerView.addItemDecoration(new NotificationsDividerDecorator(context));
        recyclerView.setAdapter(mAdapter);

        new ItemTouchHelper(new SwipeCallback()).attachToRecyclerView(recyclerView);

        mRefreshView = (SwipeRefreshLayout) root.findViewById(R.id.refresh);
        mRefreshView.setOnRefreshListener(this::refresh);

        mEmptyView = root.findViewById(android.R.id.empty);

        return root;
    }

    @Override
    public void onDestroyView() {
        mRefreshView = null;
        mEmptyView = null;
        super.onDestroyView();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_notification, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
        case R.id.action_refresh:
            refresh();
            return true;
        default:
            return super.onOptionsItemSelected(item);
        }
    }

    public void refresh() {
        mDataSubscription.set(subscribeDataSource(mAdapter.getDataSource()));
    }

    private void setEmptyViewText(CharSequence headline, CharSequence caption) {
        if (mEmptyView == null) return;
        ViewUtils.setTextOrHide(mEmptyView.findViewById(R.id.headline), headline);
        ViewUtils.setTextOrHide(mEmptyView.findViewById(R.id.caption), caption);
    }

    private Subscription subscribeDataSource(final NotificationsDataSource dataSource) {
        if (mRefreshView != null) mRefreshView.setRefreshing(true);
        if (mEmptyView != null) setEmptyViewText(null, getText(R.string.msg_fetching_content));

        return mApiClient.notifications()
                .get()
                .flatMap(Observable::from)
                .filter(notification -> notification.read_at == null)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnTerminate(() -> {
                    if (mRefreshView != null) mRefreshView.setRefreshing(false);
                })
                .subscribe(new Subscriber<Notification>() {
                    @Override
                    public void onNext(Notification notification) {
                        dataSource.addNotification(notification);
                        if (mEmptyView != null) ViewUtils.animateVisible(mEmptyView, false);
                    }

                    @Override
                    public void onCompleted() {
                        if (mEmptyView == null) return;
                        Context context = mEmptyView.getContext();
                        if (dataSource.getItemCount() == 0) {
                            setEmptyViewText(context.getText(R.string.headline_no_notifications),
                                    context.getText(R.string.msg_pull_to_refresh));
                            ViewUtils.animateVisible(mEmptyView, true);
                        } else {
                            ViewUtils.animateVisible(mEmptyView, false);
                        }
                    }

                    @Override
                    public void onError(Throwable error) {
                        Log.d("XAE", "Error fetching notifications", error);
                        if (mEmptyView != null) {
                            setEmptyViewText("Error", error.getLocalizedMessage());
                        }
                    }
                });
    }

    void markAsRead(List<Notification> notifications) {
        View container = getView();
        if (container == null) return;
        Context context = container.getContext();

        int count = notifications.size();
        String message = context.getResources()
                .getQuantityString(R.plurals.toast_notifications_read, count, count);

        final NotificationsDataSource dataSource = mAdapter.getDataSource();
        final Observable<Notification> readItems = Observable.from(notifications);

        Snackbar.make(container, message, Snackbar.LENGTH_LONG)
                .setCallback(new Snackbar.Callback() {
                    @Override
                    public void onDismissed(Snackbar snackbar, int event) {
                        if (event == DISMISS_EVENT_ACTION) return;

                        // User didn't undo the swipe, so actually mark notifications read
                        readItems.forEach(notification -> {
                            mApiClient.notifications()
                                    .markRead(notification.id)
                                    .subscribe(n -> Log.i("XAE", "notification read: " + n.id),
                                            error -> {
                                                Log.d("XAE", "markRead error", error);
                                                dataSource.addNotification(notification);
                                            });
                        });
                    }
                })
                .setAction(R.string.action_undo, v -> {
                    readItems.subscribe(dataSource::addNotification);
                })
                .show();
    }

    private class SwipeCallback extends ItemTouchHelper.SimpleCallback {

        public SwipeCallback() {
            super(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT);
        }

        @Override
        public boolean onMove(
                RecyclerView recyclerView,
                RecyclerView.ViewHolder viewHolder,
                RecyclerView.ViewHolder target) {
            return false;
        }

        @Override
        public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
            int position = viewHolder.getAdapterPosition();
            if (position == RecyclerView.NO_POSITION) return;

            List<Notification> removed = mAdapter.getDataSource().removeItem(position);
            markAsRead(removed);
        }
    }

}

File: app/src/main/java/net/xaethos/trackernotifier/fragments/NotificationsFragment.kt
package net.xaethos.trackernotifier.fragments

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.*
import net.xaethos.trackernotifier.R
import net.xaethos.trackernotifier.adapters.NotificationsAdapter
import net.xaethos.trackernotifier.adapters.NotificationsDividerDecorator
import net.xaethos.trackernotifier.api.TrackerClient
import net.xaethos.trackernotifier.models.Notification
import net.xaethos.trackernotifier.utils.Log
import net.xaethos.trackernotifier.utils.markRead
import rx.Observable
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.MultipleAssignmentSubscription

class NotificationsFragment : BaseAdapterFragment() {

    private val apiClient: TrackerClient = TrackerClient.instance
    protected override val adapter: NotificationsAdapter = NotificationsAdapter()

    private lateinit var dataSubscription: MultipleAssignmentSubscription

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        dataSubscription = MultipleAssignmentSubscription()
    }

    override fun onDestroy() {
        dataSubscription.unsubscribe()
        super.onDestroy()
    }

    override fun onCreateView(
            inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_notifications, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recyclerView?.layoutManager = LinearLayoutManager(context)
        recyclerView?.addItemDecoration(NotificationsDividerDecorator(context))
        recyclerView?.adapter = adapter
        ItemTouchHelper(SwipeCallback()).attachToRecyclerView(recyclerView)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        refresh()
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater!!.inflate(R.menu.menu_notification, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_refresh -> refresh()
            else -> return super.onOptionsItemSelected(item)
        }
        return true
    }

    override fun refresh() {
        dataSubscription.set(subscribeAdapter())
    }

    private fun subscribeAdapter(): Subscription {
        refreshView?.isRefreshing = true
        setEmptyText(R.string.msg_fetching_content, 0)

        return apiClient.notifications.get()
                .flatMap { Observable.from(it) }
                .filter { notification -> notification.read_at == null }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnTerminate { refreshView?.isRefreshing = false }
                .subscribe({ notification -> adapter.addNotification(notification) },
                        { error -> setEmptyText(getString(R.string.headline_error), error.message) },
                        { setEmptyText(R.string.headline_no_notifications, R.string.msg_pull_to_refresh) })
    }

    private fun markAsRead(notifications: List<Notification>) {
        val container = view ?: return
        val context = container.context

        val count = notifications.size
        val message = context.resources.getQuantityString(R.plurals.toast_notifications_read, count, count)

        val readItems = Observable.from(notifications)

        Snackbar.make(container, message, Snackbar.LENGTH_LONG).setCallback(object : Snackbar.Callback() {
            override fun onDismissed(snackbar: Snackbar, event: Int) {
                if (event == Snackbar.Callback.DISMISS_EVENT_ACTION) return

                // User didn't undo the swipe, so actually mark notifications read
                readItems.forEach { notification ->
                    apiClient.notifications.markRead(notification.id)
                            .subscribe(
                                    { Log.i { "notification marked read: ${it.id}" } },
                                    { error ->
                                        Log.w(error) { "markRead error" }
                                        adapter.addNotification(notification)
                                    }
                            )
                }
            }
        }).setAction(R.string.action_undo) { readItems.subscribe { adapter.addNotification(it) } }.show()
    }

    private inner class SwipeCallback : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {

        override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder): Boolean {
            return false
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
            val position = viewHolder.adapterPosition
            if (position == RecyclerView.NO_POSITION) return

            val removed = adapter.removeItem(position)
            markAsRead(removed)
        }
    }

}
